File: Algorithms and Data Structures Initial/src/Tests/TaskServiceTest.java
/*
 * Nicholas Feero
 * CS-320: Software Test & Automation
 * Module #4 Milestone | TaskServiceTest.java
 * Professor Toledo
 * February 2nd, 2025
*/

package Tests;

import static org.junit.Assert.assertTrue;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;
import java.util.ArrayList;
import org.junit.jupiter.api.DisplayName;

//creation of TaskServiceTest class
public class TaskServiceTest {
	//test creation of tasks with unique task ID using TaskService methods
	@Test
	void testTaskClass() {
		Task task = new Task("1", "Dishes", "Dirty and Clean");
		assertTrue(task.getTaskId().equals("1"));
		assertTrue(task.getTaskName().equals("Dishes"));
		assertTrue(task.getTaskDescription().equals("Dirty and Clean"));
	}
	//test to update task name by specifying task ID
	@Test
	@DisplayName("Test to update Task Name")
	void testUpdateTaskName() {
		TaskService service = new TaskService();
		Task task = new Task("11111", "Code", "Task Service File");
		service.addTask(task);
		int size = service.taskList.size();
		service.taskList.get(size - 1).getTaskId();
		service.taskList.get(size - 1).getTaskName();
		service.updateTaskName("Coding", "11111");
		service.displayTaskList();
		service.taskList.get(size - 1).getTaskName();
		assertTrue(service.taskList.get(size - 1).getTaskName().equals("Coding"));
	}
	//test to update task description by specifying task ID
	@Test
	@DisplayName("Test to update Task Description")
	void testUpdateTaskDescription() {
		TaskService service = new TaskService();
		Task task = new Task("11111", "Code", "Task Service File");
		service.addTask(task);
		int size = service.taskList.size();
		service.taskList.get(size - 1).getTaskId();
		service.taskList.get(size - 1).getTaskDescription();
		service.updateTaskDescription("Task Service Java File", "11111");
		service.displayTaskList();
		service.taskList.get(size - 1).getTaskDescription();
		assertTrue(service.taskList.get(size - 1).getTaskDescription().equals("Task Service Java File"));
	}
	//test to delete task by specifying task ID
	@Test
	@DisplayName("Test to delete task")
	void testDeleteTask() {
		TaskService service = new TaskService();
		Task task = new Task("11111", "Code", "Task Service File");
		service.addTask(task);
		service.deleteTask("11111");
		ArrayList<Task> taskListEmpty = new ArrayList<Task>();
		service.displayTaskList();
		assertEquals(service.taskList, taskListEmpty, "The task was not deleted.");
	}
	//test to add task with unique task ID
	@Test
	@DisplayName("Test to add task with unique Task ID.")
	void testAddTask() {
		TaskService service = new TaskService();
			
		Task task1 = new Task("1", "Lawn", "Mow and Trim");
		Task task2 = new Task("2", "Shopping", "Cheese, Milk, Eggs");
		Task duplicateTask1 = new Task("1", "Lawn", "Mow and Trim");
			
		assertTrue(service.addTask(task1));
		assertTrue(service.addTask(task2));
			
		assertFalse(service.addTask(duplicateTask1));
	}
}

File: Algorithms and Data Structures Enhancement/TaskServiceTest.cpp
#include "gtest/gtest.h"
#include "TaskService.h"
#include "Task.h"
#include <memory>
#include <vector>

// Test creation of tasks with unique task ID using TaskService methods
TEST(TaskServiceTests, ValidTaskCreation) {
    Task t("1", "Dishes", "Dirty and Clean");
    EXPECT_EQ(t.getTaskId(), "1");
    EXPECT_EQ(t.getTaskName(), "Dishes");
    EXPECT_EQ(t.getTaskDescription(), "Dirty and Clean");
}

// Test to update task name by specifying task ID
TEST(TaskServiceTests, UpdateTaskName) {
    TaskService service;
    auto task = std::make_shared<Task>("11111", "Code", "Task Service File");
    ASSERT_TRUE(service.addTask(task));
    size_t size = service.taskList.size();
    service.updateTaskName("Coding", "11111");
    ASSERT_EQ(service.taskList[size - 1]->getTaskName(), "Coding");
}

// Test to update task description by specifying task ID
TEST(TaskServiceTests, UpdateTaskDescription) {
    TaskService service;
    auto task = std::make_shared<Task>("11111", "Code", "Task Service File");
    ASSERT_TRUE(service.addTask(task));
    size_t size = service.taskList.size();
    service.updateTaskDescription("Task Service C++ File", "11111");
    ASSERT_EQ(service.taskList[size - 1]->getTaskDescription(), "Task Service C++ File");
}

// Test to delete task by specifying task ID
TEST(TaskServiceTests, DeleteTask) {
    TaskService service;
    auto task = std::make_shared<Task>("11111", "Code", "Task Service File");
    ASSERT_TRUE(service.addTask(task));
    service.deleteTask("11111");
    std::vector<std::shared_ptr<Task>> emptyVec;
    EXPECT_EQ(service.taskList, emptyVec);
}

// Test to add task with unique task ID
TEST(TaskServiceTests, AddTaskWithUniqueId) {
    TaskService service;
    auto task1 = std::make_shared<Task>("1", "Lawn", "Mow and Trim");
    auto task2 = std::make_shared<Task>("2", "Shopping", "Cheese, Milk, Eggs");
    auto duplicateTask1 = std::make_shared<Task>("1", "Lawn", "Mow and Trim");

    EXPECT_TRUE(service.addTask(task1));
    EXPECT_TRUE(service.addTask(task2));
    EXPECT_FALSE(service.addTask(duplicateTask1));
}
